tutorials/requirements.txt

```diff
@@ -0,0 +1,4 @@
+numpy
```

Comment:
there's an environment.yml in the root directory - I don't have a strong pref for conda or pip, but let's pick only one.

Reply:
If we need to pick one then pip is probably more commonly used. But perhaps we could make an effort for the sake of inclusivity and support both camps? Our `environment.yml` and `requirements.txt` are so simple that we can easily make them equivalent.

What worries me more is the following: one session or another will require a really obscure module/library, and one way to isolate it from the rest is to have a dedicated `environment.yml` / `requirements.txt` - how can we deal with such scenario?

Reply:
Well, one way to deal with such scenarios is to have a standard `environment.yml`/`requirements.txt` for most days that is used by the default GitHub action. Days with specific requirements have `environment.yml`/`requirements.txt`  at the folder level, are checked with dedicated action, and are excluded from the default action. A bit contorted but may possibly work.

Reply:
One issue with pip is it doesn't specify which version of Python to use. Presumably a later python could change behaviour and break the notebooks. If you're going to pick requirements.txt, then add runtime.txt as well: https://repo2docker.readthedocs.io/en/latest/config_files.html#runtime-txt-specifying-runtimes. 

I think it doesn't make a lot of sense to have both. If you look at the behaviour of repo2docker (and therefore binder), it basically starts by installing via conda environment.yml and then installing requirements.txt on top of that. So if they specify the same packages then the packages get installed twice.

I'm ok with folder-specific requirements.txt for special requirements.

Reply:
I was not very clear in my earlier message - sorry! My proposal is to provide both `environment.yml` and `requirements.txt` for people to run notebooks outside colab environment. These would contain the same modules/libraries (with `environment.yml` additionally specifying Python version).

In the `README.md` we would specify the version(s) of Python we support and are checked with CI. Ideally, these files are generated from the colab environment, and I suspect we don't need to be specific about library versions and python versions are checked by CI.

This way, we would seamlessly support the following platforms:
- colab notebooks
- jupyter notebooks on binderhub
(if binderhub always executes `requirements.txt` in the root of the repo after `environment.yml`, then we would place the reference `requirements.txt` and `environment.yml` elsewhere (under `tutorials`?)
- jupyter anaconda with repo cloning followed by `environment.yml` (eventually on a virtual environment and we provide instructions to do so)
- jupyter standard with repo cloning followed by `requirements.txt`, eventually on a virtual environment and we provide instructions to do so)

This is what I meant by inclusivity and supporting both camps - we provide reference `environment.yml` and `requirements.txt` to facilitate running notebooks locally, which a lot of users will do during or after the this year's academy.

Reply:
Of course, we could also look into converting the CI script to conda - will definitely work.